### DISCUSSION TITLE
is it possible to replicate a conan datasource?

### DISCUSSION THREAD

**Discussion by jjcaballero** [renovatebot/renovate#8411] [Request Help]
Hi all,

There is already an open issue (https://github.com/renovatebot/renovate/issues/5933) requesting the support of conan packages. I would really love to count on this feature in the future :-)

In the meantime, I was wondering if I could workaround this through some regex expressions or if a specific datasource could be used to accomplish the detection of new generic versions. All in all, what I would need is a datasource configuration that would detect new folders in a given registry considering the given versioning, thus I was hoping this could be done. As I am new in Renovate, I am probably missing some pieces, sorry beforehand.

This is what I am trying so far:

File version with a single dependency in the root directory:
`single_dependency.txt`

Content of the version file: 
`1.0.0.0`

Registry:
```
https://internal-artifactory.com/artifactory/folder/
                                              |------ DEPENDENCY/
                                                        |---------- 1.0.0.0/...
                                                        |---------- 1.0.0.1/...
                                                        |---------- 1.0.1.0/...

```

And renovate.json:

```
{
    "regexManagers": [
        {
            "fileMatch": ["^single_dependency.txt$"],
            "matchStrings": ["^(?<currentValue>.*?)$"],
            "depNameTemplate": "DEPENDENCY",
            "datasourceTemplate": "pypi"
        }
    ],

    "packageRules": [
        {
            "datasources": ["pypi"],
            "registryUrls": ["https://internal-artifactory.com/artifactory/folder/"],
            "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(\\.(?<compatibility>.*))?$"
        }
    ],

    "enabledManagers": ["regex"]
}
```

With this, I am always getting: Found 0 package file(s)

```
DEBUG: Using file match: ^single_dependency.txt$ for manager regex (repository=...)
DEBUG: Matched 1 file(s) for manager regex: single_dependency.txt (repository=...)
DEBUG: Found 0 package file(s) (repository=...)
```

I have tried with pypi, docker, git-refs and maven. Of course, the package under the version folders won't match the structure of pip, docker, etc. Not sure if that is the root cause. Another source of problems could be the 4 digits version. 

Please, any suggestion & feedback is more than welcome. Thanks!

**Comment by rarkins** [comment]
If you're seeing `DEBUG: Found 0 package file(s) (repository=...)` then I think it means that the regex is not working properly to extract any dependencies out.

Can you create an example repository of what you're trying with?

BTW I'm not sure you'll have any luck with the datasource as the next step, but this first extraction step seems to be what's failing.

**Reply by jjcaballero** [reply]
Thank @rarkins. 
I hope this can work as an example: https://github.com/jjcaballero/renovate-example-for-discussion-8411/blob/main/renovate.json

**Reply by rarkins** [reply]
What's the full debug log with that repo currently?

**Reply by jjcaballero** [reply]
Sorry, I didn't have renovate configured in the example. Now, it is and thanks to this I found an issue in matchStrings parameter (already fixed). The result still shows _Found 0 package file(s)_ though.

```
INFO: Repository started
{
  "renovateVersion": "24.28.1"
}
DEBUG: Using localDir: /mnt/renovate/gh/jjcaballero/renovate-example-for-discussion-8411
DEBUG: Repository cache not found
{
  "cacheFileName": "/tmp/renovate-cache/renovate/repository/github/jjcaballero/renovate-example-for-discussion-8411.json"
}
DEBUG: initRepo("jjcaballero/renovate-example-for-discussion-8411")
DEBUG: Overriding default GitHub endpoint
{
  "endpoint": "https://api.github.com/"
}
DEBUG: No dangling containers to remove
DEBUG: jjcaballero/renovate-example-for-discussion-8411 default branch = main
DEBUG: Using personal access token for git init
DEBUG: resetMemCache()
DEBUG: checkOnboarding()
DEBUG: isOnboarded()
DEBUG: findFile(renovate.json)
DEBUG: Initializing git repository into /mnt/renovate/gh/jjcaballero/renovate-example-for-discussion-8411
DEBUG: git clone completed
{
  "durationMs": 591
}
DEBUG: latest commit
{
  "latestCommitDate": "2021-01-26T12:09:45+01:00"
}
DEBUG: Setting git author name
{
  "gitAuthorName": "Renovate Bot"
}
DEBUG: Setting git author email
{
  "gitAuthorEmail": "bot@renovateapp.com"
}
DEBUG: Config file exists
{
  "fileName": "renovate.json"
}
DEBUG: Retrieving issueList
DEBUG: Retrieved 1 issues
DEBUG: Repo is onboarded
DEBUG: Found renovate.json config file
DEBUG: Repository config
{
  "fileName": "renovate.json",
  "config": {
    "baseBranches": [
      "main"
    ],
    "regexManagers": [
      {
        "fileMatch": [
          "^alien/version.txt$"
        ],
        "matchStrings": [
          "^(?<currentValue>.+)$"
        ],
        "depNameTemplate": "DEPENDENCY",
        "datasourceTemplate": "docker",
        "versioningTemplate": "^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "packageRules": [
      {
        "datasources": [
          "docker"
        ],
        "registryUrls": [
          "https://internal-artifactory.com/artifactory/folder"
        ],
        "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "enabledManagers": [
      "regex"
    ]
  }
}
DEBUG: migrateAndValidate()
DEBUG: No config migration necessary
DEBUG: massaged config
{
  "config": {
    "baseBranches": [
      "main"
    ],
    "regexManagers": [
      {
        "fileMatch": [
          "^alien/version.txt$"
        ],
        "matchStrings": [
          "^(?<currentValue>.+)$"
        ],
        "depNameTemplate": "DEPENDENCY",
        "datasourceTemplate": "docker",
        "versioningTemplate": "^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "packageRules": [
      {
        "datasources": [
          "docker"
        ],
        "registryUrls": [
          "https://internal-artifactory.com/artifactory/folder"
        ],
        "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "enabledManagers": [
      "regex"
    ],
    "extends": [
      "github>whitesource/merge-confidence:beta"
    ]
  }
}
DEBUG: migrated config
{
  "config": {
    "baseBranches": [
      "main"
    ],
    "regexManagers": [
      {
        "fileMatch": [
          "^alien/version.txt$"
        ],
        "matchStrings": [
          "^(?<currentValue>.+)$"
        ],
        "depNameTemplate": "DEPENDENCY",
        "datasourceTemplate": "docker",
        "versioningTemplate": "^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "packageRules": [
      {
        "datasources": [
          "docker"
        ],
        "registryUrls": [
          "https://internal-artifactory.com/artifactory/folder"
        ],
        "versioning": "regex:^(?<major>\\d+)(\\.(?<minor>\\d+))+(\\.(?<patch>\\d+))+(\\.(?<compatibility>.*))+$"
      }
    ],
    "enabledManagers": [
      "regex"
    ],
    "extends": [
      "github>whitesource/merge-confidence:beta"
    ]
  }
}
DEBUG: Found repo ignorePaths
{
  "ignorePaths": [
    "**/node_modules/**",
    "**/bower_components/**"
  ]
}
DEBUG: detectSemanticCommits()
DEBUG: getCommitMessages
DEBUG: Semantic commits detection: angular
DEBUG: angular semantic commits detected
DEBUG: Setting branchPrefix: renovate/
DEBUG: Cannot read vulnerability alerts
DEBUG: No vulnerability alerts found
DEBUG: processRepo()
DEBUG: baseBranches
{
  "baseBranches": [
    "main"
  ]
}
DEBUG: baseBranch: main
DEBUG: extract()
DEBUG: Setting current branch to main
DEBUG: latest commit
{
  "branchName": "main",
  "latestCommitDate": "2021-01-26T12:09:45+01:00"
}
DEBUG: Applying enabledManagers filtering
DEBUG: Using file match: ^alien/version.txt$ for manager regex
DEBUG: Matched 1 file(s) for manager regex: alien/version.txt
DEBUG: Found 0 package file(s)
INFO: Dependency extraction complete
{
  "baseBranch": "main",
  "stats": {
    "managers": {},
    "total": {
      "fileCount": 0,
      "depCount": 0
    }
  }
}
DEBUG: baseBranch: main
DEBUG: Package releases lookups complete
{
  "baseBranch": "main"
}
DEBUG: packageFiles with updates
{
  "config": {}
}
DEBUG: branchifyUpgrades
DEBUG: 0 flattened updates found:
DEBUG: Returning 0 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: processRepo()
DEBUG: Processing 0 branches:
DEBUG: Calculated maximum PRs remaining this run
{
  "prsRemaining": 99
}
DEBUG: PullRequests limit = 99
DEBUG: Calculated maximum branches remaining this run
{
  "branchesRemaining": 99
}
DEBUG: Branches limit = 99
DEBUG: Removing any stale branches
DEBUG: config.repoIsOnboarded=true
DEBUG: No renovate branches found
DEBUG: closeIssue(1)
DEBUG: Issue closed
{
  "number": 1
}
DEBUG: Repository timing splits (milliseconds)
{
  "splits": {
    "init": 3685,
    "extract": 1436,
    "lookup": 1,
    "update": 1
  },
  "total": 5960
}
DEBUG: http statistics
{
  "hostStats": [
    "api.github.com, 5 requests, 294ms request average, 0ms queue average"
  ],
  "totalRequests": 5
}
INFO: Repository finished
{
  "durationMs": 5960
}
```

notifying: @rarkins (in case I should have, if not, sorry for the noise)

**Reply by jjcaballero** [reply]
Found it. The 4-digits-dependency is extracted now. 
The problem was _matchStrings_ from _regexManagers_ does not seem to process the beginning and end of line characters (^ and $), without those it is working.

Now, the error is looking up the dependency in the datasource, as expected because the registry does not exist in this test bench:
```
DEBUG: Datasource connection error
{
  "datasource": "pypi",
  "lookupName": "DEPENDENCY",
  "errCode": "ENOTFOUND"
}
```
I will try different approaches in my original test bench where I have the registry in place.

**Reply by rarkins** [reply]
Great that you progressed. Does Conan have its own dedicated registry/datasource, or does it usually reference things like GitHub repo branches or tags?

**Reply by jjcaballero** [reply]
Sorry, I am not sure if I fully understand your question. In my company, the conan packages are stored in Artifactory, like many other packages (docker, maven, generic artifacts, etc). From my humble view: every version of conan package has a conanfile.py in it which defines how the package will be installed. I would say that is the most meaningful file/difference in conan packages.

FWIW, as long as there was a _datasource_ which allowed me to detect new folders in a registry comparing the version, that would do it the trick. I have tried _gradle-version_ _datasource_ because it seemed simpler than others. And I am getting _Response code 406 (Not Acceptable)_ because the request has some constraints in the package to download.

Is there any documentation for developers wrt. _datasources_? it would be great if I can understand better how they work.
Thanks, @rarkins

**Reply by rarkins** [reply]
There's some documentation for datasources in the README here: https://github.com/renovatebot/renovate/tree/master/lib/datasource

And naturally each of the subfolders there is an example which can be copied.

Have you found a specification for the Conan server API?

**Reply by jjcaballero** [reply]
@rarkins I am afraid not. The following link can work to have an idea of how Conan packages are stored in Artifactory: [jfrog - Artifactory: Conan Repositories](https://www.jfrog.com/confluence/display/JFROG/Conan+Repositories) (section _Viewing Individual Conan Package Information_)
But I haven't found anything related to the specification you asked for in the [official documentation](https://docs.conan.io/en/latest/official).

Is there a way to create a docker image similar to the [official image](https://hub.docker.com/r/renovate/renovate/) from a fork? We run renovate in GitLab, I was wondering if I could test directly the changes in my fork with our Artifactory registry through a previously built Docker image, instead of using the official one. I see there is a _./lib/manager/regex/__fixtures__/Dockerfile_ with renovate.js as the entrypoint. If I replaced the renovate.js with the one built in my fork under _./dist/renovate.js_, would this work? I hope so.

I would need to figure out (ideally soon) if this conan as datasource is something I can do or not and I think if I was able to create such a test docker image from my fork, I would have everything I need. Any official contribution would take more time to implement for sure. Let's say, now we are in the design stage and I need to decide if we go with Renovate or something else. We use Renovate for docker, pypi and others, I would love to have Renovate working for our projects with conan dependencies too.

Thanks!

**Reply by rarkins** [reply]
Yes, you should be able to rebuild and replace it. I'd recommend running it locally from source if you can though (e.g. assuming you have access to Artifactory from your laptop).

BTW I looked into it a little myself yesterday and it seems like the Conan server API is "the Artifactory API", which I've got to say is a bit weird if so. We of course don't need to support the whole protocol though - usually just enough to get the list of versions per-package.

**Reply by jjcaballero** [reply]
Unbelievable, I am already running it locally using our GitLab instance. Awesome! 
(great instructions https://docs.renovatebot.com/self-hosting/ and greater hint to run it locally). Thanks!
I am sure it will take time until I manage to create a humble new datasource :-) , I will keep you post.

**Comment by jjcaballero** [comment]
Hi @rarkins,
I managed to read some versions in a very humble conan datasource prototype which just reads the folders where the user points to in the registry.
My problem now is comparing the 4 digits versions. 
In [lib/versioning/loose/generic.ts](https://github.com/renovatebot/renovate/blob/master/lib/versioning/loose/generic.ts), the 4 digits versions are considered equals when the 3 first digits are the same as they are the same Semver version. (as mentioned in the [readme](https://github.com/renovatebot/renovate/tree/master/lib/versioning/loose).

Is there a way to work around this? Any suggestion? Is the only way to create a new versioning scheme?

Thanks

**Reply by rarkins** [reply]
Try "loose" versioning

**Comment by jjcaballero** [comment]
I am bringing back this topic because I have been allowed by my company (Nuance Communications) to spend some time on this.

Implementing the Conan manager mentioned in [issue 5933](https://github.com/renovatebot/renovate/issues/5933) might be too challenging for me as my first contribution, I was hoping there might be other alternatives (e.g.: focus on the datasource part).

In our use case, we are using a Renovate fork to update conan dependencies. We use: regex manager, pep440 versioning and the humble datasource I implemented. 

This datasource is just getting all folders in the url composed by _registryUrl_ and _lookupName_. To do this, the Http class defined in _util/http/index.ts_ is used. The gathered folders in the resulting url are the release candidates. So, for each found folder, a Release is created. The version of the Release is the name of the folder. Finally, the getReleases returns this list of Releases.

To me, this is like a generic Jfrog Artifactory datasource (or probably a simply generic datasource). Jfrog Artifactory is the registry we are using, but probably it is pretty similar to Bintray.

I would love to hear maintainers' thoughts about this idea and hopefully, define the next steps.

Thanks!

**Reply by rarkins** [reply]
If you're using a "generic" JFrog Artifactory datasource which you think has wider applicability (either for other package managers or other users) then it would definitely be a good idea to add it as a new datasource. Naming could be something like `artifactory-generic` for example. Otherwise if it was something specific to Conan then you'd want to call it `conan` or `artifactory-conan`.

Perhaps best to start a new feature request issue for Conan *datasource* (separate from the existing one from Conan manager) and we can help you there. Also feel free to email me to get an invite to our developer Slack channel, which you might find easier to ask rapid questions in (although we try to be pretty responsive in GitHub and ultimately prefer that technical decisions are documented in GitHub too).

**Reply by jjcaballero** [reply]
For reference: working on it: https://github.com/renovatebot/renovate/issues/11474
